ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/materialization/ks/WindowStoreCacheBypassTest.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.QueryableStoreType;
+import org.apache.kafka.streams.state.ReadOnlyWindowStore;
+import org.apache.kafka.streams.state.StateSerdes;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.apache.kafka.streams.state.WindowStore;
+import org.apache.kafka.streams.state.WindowStoreIterator;
+import org.apache.kafka.streams.state.internals.CompositeReadOnlyWindowStore;
+import org.apache.kafka.streams.state.internals.MeteredWindowStore;
+import org.apache.kafka.streams.state.internals.StateStoreProvider;
+import org.apache.kafka.streams.state.internals.WrappedStateStore;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class WindowStoreCacheBypassTest {
+
+  private static final Schema SCHEMA = SchemaBuilder.struct().field("a", SchemaBuilder.int32());
+  private static final Struct SOME_KEY = new Struct(SCHEMA).put("a", 1);
+  private static final byte[] BYTES = new byte[] {'a', 'b'};
+
+  @Mock
+  private QueryableStoreType<ReadOnlyWindowStore<Struct, ValueAndTimestamp<GenericRow>>>
+      queryableStoreType;
+  @Mock
+  private StateStoreProvider provider;
+  @Mock
+  private MeteredWindowStore<Struct, ValueAndTimestamp<GenericRow>> meteredWindowStore;
+  @Mock
+  private WindowStore<Bytes, byte[]> windowStore;
+  @Mock
+  private WrappedWindowStore<Bytes, byte[]> wrappedWindowStore;
+  @Mock
+  private StateStore stateStore;
+  @Mock
+  private WindowStoreIterator<byte[]> windowStoreIterator;
+  @Mock
+  private StateSerdes<Struct, ValueAndTimestamp<GenericRow>> serdes;
+
+  private CompositeReadOnlyWindowStore<Struct, ValueAndTimestamp<GenericRow>> store;
+  private WrappedStateStore<StateStore, Struct, ValueAndTimestamp<GenericRow>> wrappedStateStore;
+
+  @Before
+  public void setUp() {
+    store = new CompositeReadOnlyWindowStore<>(provider, queryableStoreType, "foo");
+  }
+
+  @Test
+  public void shouldCallUnderlyingStore() throws IllegalAccessException {
+    when(provider.stores(any(), any())).thenReturn(ImmutableList.of(meteredWindowStore));
+    SERDES_FIELD.set(meteredWindowStore, serdes);
+    when(serdes.rawKey(any())).thenReturn(BYTES);
+    when(meteredWindowStore.wrapped()).thenReturn(wrappedWindowStore);
+    when(wrappedWindowStore.wrapped()).thenReturn(windowStore);
+    when(windowStore.fetch(any(), any(), any())).thenReturn(windowStoreIterator);
+    when(windowStoreIterator.hasNext()).thenReturn(false);
+
+    WindowStoreCacheBypass.fetch(store, SOME_KEY, Instant.ofEpochMilli(100), Instant.ofEpochMilli(200));
+    verify(windowStore).fetch(new Bytes(BYTES), Instant.ofEpochMilli(100L), Instant.ofEpochMilli(200L));
+  }
+
+  @Test
+  public void shouldAvoidNonWindowStore() throws IllegalAccessException {
```

Comment:
It looks like this test verifies that we bottom out at a wrapped store when it doesn't wrap a WindowStore, right?
It seems like a similar test could positively verify that we actually do skip the caching layer, or any other wrapped layer, but I didn't see that test. Did I miss it?

Reply:
The logic was technically cache agnostic.  So long as the last WindowStore layer is not a cache, which we know it wouldn't be since the caching layer is built upon that, this check would hopefully be sufficient.

I changed the bypassing logic so that it verifies that it actually passes the caching layer when it's run and the tests verify that as well.

Reply:
Actually, I removed the bypass check because it's possible that the user can configure it without a cache.  I don't really want to try to reproduce the logic for determining if cache is enabled, so I removed this bypass check.